src/lib.rs

```diff
@@ -298,19 +295,50 @@ where
         })?;
         writer.write(XmlEvent::start_element("graphml").attr("xmlns", NAMESPACE_URL))?;
 
-        // emit graph with nodes/edges and possibly weights
-        self.emit_graph(writer, &mut attributes)?;
+        // First pass to extract keys
+        let attributes = self.extract_attributes();
```

Comment:
You can further clean up the code by merging `extract_attributes` and `emit_keys`. Since the `attributes` variable is not used anywhere else it is not really necessary to even create it in the first place.